File: backend/app/schemas/analysis.py
"""
Pydantic schemas for analysis endpoints.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class AnalysisCreateRequest(BaseModel):
    """Request schema for creating an analysis job."""
    baseline_s3_key: str = Field(..., description="S3 key for baseline/current policy PDF")
    renewal_s3_key: str = Field(..., description="S3 key for renewal policy PDF")
    metadata_company_name: Optional[str] = Field(None, description="Company name (optional)")
    metadata_policy_type: Optional[str] = Field(None, description="Policy type (optional)")


class AnalysisJobResponse(BaseModel):
    """Response schema for analysis job (status)."""
    job_id: str
    status: str  # pending, processing, completed, failed
    created_at: datetime
    updated_at: datetime
    baseline_filename: str
    renewal_filename: str
    progress: Optional[int] = None
    message: Optional[str] = None
    estimated_completion_time: Optional[str] = None
    error_message: Optional[str] = None
    
    class Config:
        from_attributes = True


class CoverageChange(BaseModel):
    """Schema for a single coverage change."""
    id: Optional[str] = None
    category: str
    change_type: str
    title: str
    description: str
    baseline_value: str
    renewal_value: str
    change_amount: Optional[str] = None
    percentage_change: Optional[float] = None
    confidence: Optional[float] = None
    page_references: Optional[Dict[str, List[int]]] = None


class PremiumComparison(BaseModel):
    """Schema for premium comparison."""
    baseline_premium: Optional[float] = None
    renewal_premium: Optional[float] = None
    difference: Optional[float] = None
    percentage_change: Optional[float] = None


class AnalysisSummary(BaseModel):
    """Schema for analysis summary."""
    total_changes: int
    change_categories: Dict[str, int]


class AnalysisResultResponse(BaseModel):
    """Response schema for completed analysis results."""
    job_id: str
    status: str
    summary: AnalysisSummary
    changes: List[CoverageChange]
    premium_comparison: PremiumComparison
    suggested_actions: List[Dict[str, str]]
    educational_insights: List[Dict[str, str]]
    metadata: Dict[str, Any]
    
    class Config:
        from_attributes = True


class AnalysisListItem(BaseModel):
    """Schema for analysis list item."""
    job_id: str
    status: str
    created_at: datetime
    completed_at: Optional[datetime] = None
    baseline_filename: str
    renewal_filename: str
    total_changes: Optional[int] = None
    company_name: Optional[str] = None
    
    class Config:
        from_attributes = True

File: frontend/src/types/analysis.ts
export type ChangeType = "increased" | "decreased" | "modified" | "added" | "removed";

export type ChangeCategory = 
  | "coverage_limit" 
  | "deductible" 
  | "exclusion" 
  | "premium" 
  | "terms_conditions"
  | "other";

export interface PolicyChange {
  id: string;
  category: ChangeCategory;
  change_type: ChangeType;
  title: string;
  description: string;
  baseline_value: string;
  renewal_value: string;
  change_amount?: string;
  percentage_change?: number;
  confidence: number;
  page_references: {
    baseline: number[];
    renewal: number[];
  };
}

export interface SuggestedAction {
  category: ChangeCategory;
  action: string;
  educational_context: string;
}

export interface EducationalInsight {
  change_type: string;
  insight: string;
}

export interface PremiumComparison {
  baseline_premium: number | null;
  renewal_premium: number | null;
  difference: number | null;
  percentage_change: number | null;
}

export interface AnalysisSummary {
  total_changes: number;
  change_categories: Record<ChangeCategory, number>;
}

export interface AnalysisResult {
  job_id: string;
  status: "processing" | "completed" | "failed";
  summary: AnalysisSummary;
  changes: PolicyChange[];
  premium_comparison: PremiumComparison;
  suggested_actions: SuggestedAction[];
  educational_insights: EducationalInsight[];
  metadata: {
    analysis_version: string;
    model_version: string;
    processing_time_seconds: number;
    completed_at: string;
  };
}

export interface JobStatus {
  job_id: string;
  status: "processing" | "completed" | "failed";
  progress: number;
  message: string;
  updated_at: string;
}

export interface UploadSession {
  upload_urls: Array<{
    url: string;
    key: string;
    expires_at: string;
  }>;
}